### PR TITLE
fix for visualization of surface corners behind the camera

### PR DESCRIPTION
**Pull request by s-westphal**
Hi!
the opencv `perspectiveTransform` function maps point back to euclidean space independent of the z-coordinate, which causes mirrored x and y coordinates for negative z. So this fix takes the absolute value of z for the conversion from homogeneous to euclidean coordinates. 
Feel free to merge or make changes!

### CODE REVIEW

**Comment by romanroibu** [comment]
Hey @marc-tonsen,

Could you please review this and let us know what you think? Thanks!

File: src/surface_tracker/location.py [hunk 1]
```diff
@@ -298,6 +298,17 @@ def __map_points(
 ##### Helper functions
 
 
+def _perspective_transform(points, transform_matrix):
```
Comment (euryalus) [inline src/surface_tracker/location.py]: I think, you can get rid of the for loop by defining this function (I think this is also changed later in this PR, right?) as:

```
def _perspective_transform(points, transform_matrix):
    homogeneous_points = cv2.convertPointsToHomogeneous(points)[:,0,:]
    projected_points = (transform_matrix @ homogeneous_points.T).T
    projected_points[:,2] = np.abs(projected_points[:,2])
    return cv2.convertPointsFromHomogeneous(projected_points)[:,0,:]

```

File: src/surface_tracker/location.py [hunk 2]
```diff
@@ -298,15 +303,63 @@ def __map_points(
...
+    return all(
+        (
+            is_clockwise_triangle((p0, p1, p2)) == clockwise,
+            is_clockwise_triangle((p1, p2, p3)) == clockwise,
+            is_clockwise_triangle((p2, p3, p0)) == clockwise,
+        )
+    )
+
+
+def project_points_pos_z(homogeneous_points, transform_matrix):
```
Comment (euryalus) [inline src/surface_tracker/location.py:324]: Also here, I think you can get rid of the for loop. Probably known to you, but for an (n,m) array of points you can calculate the row-wise norm by:
`np.linalg.norm(points, axis=1, keepdims=True)`

**Review by euryalus — APPROVED** [review]

**Review by marc-tonsen — APPROVED** [review]
I can't really judge the correctness of these changes anymore, but in terms of style they look good to me!

File: src/surface_tracker/location.py (hunk 2 above)
Comment (s-westphal) [reply src/surface_tracker/location.py:324]: I don't see how this applies here, as only the norm of the x and y coordinates is computed if the z value is negative. (This ensures that points lying in negative z direction are not mirrored at the origin, when converting from homogeneous coordinates to euclidean space, and at the same time are far away and not visible in the image.)

Comment (euryalus) [reply src/surface_tracker/location.py:324]: Yeah, I see. I thought of something like this:
```
homogeneous_points = cv2.convertPointsToHomogeneous(points)[:,0,:]
projected_points = (transform_matrix @ homogeneous_points.T).T
idx = projected_points[:,2]<0
temp = np.vstack((-projected_points[idx,2], np.linalg.norm(projected_points[idx,:2], axis=1))
projected_points[idx, 2] = np.min(temp, axis=0)
```
But then I agree that does not make it more readable and speed is not an issue here I suppose.

**Pull request merged**